CI/OpenBSD: functional tests

vext01 opened the pull request:
This should run the functional tests now.

I expect a lot of failures and maybe a hang. Let's see what happens.

FAO @justinmk

vext01 committed:
CI/OpenBSD: run functional tests

Adapt some tests for OpenBSD:

- scrollback_spec:
  - seq(1) is not available on OpenBSD: we'd use jot(1).
  - Instead use a (hopefully) portable awk(1) snippet.
- channels_spec
- job_spec
- tui_spec

justinmk commented on test/functional/terminal/scrollback_spec.lua:
```diff
@@ -448,7 +448,7 @@ describe("'scrollback' option", function()
     else
-      feed_data('for i in $(seq 1 40); do echo "line$i"; done\n')
+      feed_data('awk "BEGIN{for(n=1;n<=40;n++) print \\\"line\\\" n}"\n')
     end
```
i think we can use `shell-test` now, for all platforms including Windows. 

Example: https://github.com/neovim/neovim/blob/6b45e12d67ec01c8a/test/functional/ui/searchhl_spec.lua#L150

    feed([[:terminal "]]..nvim_dir..[[/shell-test" REP 5000 foo<cr>]])

vext01 replied:
```
Expected (anywhere): "line30                        " 
Actual:                             
  |test" REP 30 line           < |                                                                        
  |/bin/sh: :terminal: not found |                                                                        
  |myhost$ {1: }                   |                                                                    
  |                              |                   
  |                              |
  |                              |                                                                        
  |{3:-- TERMINAL --}                |
```

It might be easier for you to do this part, as I don't know anything about lua or your test suite ;)

vext01 replied:
I also tried killing the `:terminal` prefix, but it looks to me like we are expecting to execute a ex mode command, but instead pass it to the shell.

vext01 replied:
Also noticed this stuff in that same test file:
```
  local function set_fake_shell()      
    -- shell-test.c is a fake shell that prints its arguments and exits.
    nvim('set_option', 'shell', nvim_dir..'/shell-test')
    nvim('set_option', 'shellcmdflag', 'EXE')          
  end
```

Not sure if that interferes?

justinmk commented:
this test is a known issue (failes on travis ubuntu sometimes):

```
[  ERROR   ] test/functional/ui/cmdline_spec.lua @ 830: cmdline redraw with <Cmd>
test/functional/ui/screen.lua:580: Row 2 did not match.
```

vext01 commented:
Well I made a little progress. Most of these are just attempts to use `seq` on BSD. I've fixed them using either `awk` or `jot` for now, but I hope @justinmk can find a better way.

[current test run seems to be hanging]

justinmk commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -791,7 +791,9 @@ describe("TUI 'term' option", function()
+    if string.lower(uname()) == "openbsd" then
+      assert_term("xterm", "xterm")
+    elseif is_bsd then  -- BSD lacks terminfo, builtin is always used.
       assert_term("xterm", "builtin_xterm")
```
does openbsd have terminfo ?

justinmk commented:
```
[ RUN      ] swapfile detection always show swapfile dialog #8840 #9027: ERR
test/functional/helpers.lua:240: test/functional/ui/screen.lua:898: bad argument #1 to 'unpack' (table expected, got number)

stack traceback:
	test/functional/helpers.lua:240: in function 'run_session'
	test/functional/ui/screen.lua:527: in function '_wait'
	test/functional/ui/screen.lua:368: in function 'expect'
	test/functional/ex_cmds/swapfile_preserve_recover_spec.lua:135: in function <test/functional/ex_cmds/swapfile_preserve_recover_spec.lua:102>
```

vext01 commented:
It does, yeah, but I guess it behaves differently than expected?

https://man.openbsd.org/terminfo.5

On Sun, 1 Sep 2019, 19:18 Justin M. Keyes, <notifications@github.com> wrote:

> *@justinmk* commented on this pull request.
> ------------------------------
>
> In test/functional/terminal/tui_spec.lua
> <https://github.com/neovim/neovim/pull/10804#discussion_r319769927>:
>
> > @@ -791,7 +791,9 @@ describe("TUI 'term' option", function()
>    end)
>
>    it('gets system-provided term if $TERM is valid', function()
> -    if is_bsd then  -- BSD lacks terminfo, builtin is always used.
> +    if string.lower(uname()) == "openbsd" then
> +      assert_term("xterm", "xterm")
> +    elseif is_bsd then  -- BSD lacks terminfo, builtin is always used.
>        assert_term("xterm", "builtin_xterm")
>
> does openbsd have terminfo ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/neovim/neovim/pull/10804?email_source=notifications&email_token=AAETWG7RIG7NJY4TDNVYMW3QHQBP7A5CNFSM4IMSWWOKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCDKGN3I#pullrequestreview-282355437>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAETWG5FUMQWTKPCHZHBZTDQHQBP7ANCNFSM4IMSWWOA>
> .
>

justinmk committed:
test: "can have two timers": retry()

ref #10768

justinmk committed:
screen.lua: dump payload on handler failure

For debugging failures like:

    test/functional/helpers.lua:240: test/functional/ui/screen.lua:898:
    bad argument #1 to 'unpack' (table expected, got number)

    test/functional/helpers.lua:240: test/functional/ui/screen.lua:708:
    attempt to index local 'item' (a number value)

ref #10804

justinmk committed:
test: shell-test.c: flush all streams

justinmk commented:
Seems like 5487624730b60426632326dde41383069754916b is still a problem on OpenBSD, or maybe it's a symptom of something more fundamental:  some of the builds  here failed on [bad redraw-event](https://github.com/neovim/neovim/pull/10804#issuecomment-526943159). Thus 65776ed was added for better visibility, but then the failures mostly stopped...

justinmk committed:
test/OpenBSD: skip some tests

Temporary workaround to unblock CI for OpenBSD.

justinmk committed:
test/uname(): always lowercase

justinmk commented on .builds/openbsd.yml:
```diff
@@ -28,11 +28,14 @@ tasks:
-    cd neovim
+    # functional tests
+    cd neovim/build
+    # cmake --build . --config Debug --target functionaltest
```
Disabled on master until we can fix the remaining issues. In particular we need `ci/common/test.sh:check_logs()` (which checks for core dumps and prints the backtrace with gdb).  Or we could ssh into sourcehut and get the backtrace manually.
